db/migrate/20180410071554_current_policy.rb

```diff
@@ -0,0 +1,58 @@
+Sequel.migration do
+  up do
+    alter_table :policy_versions do
+      set_column_type :created_at, :timestamptz
+      add_column :commited_at, :timestamptz,
+        index: true, # to quickly find the null
+        null: true # will be enforced by a trigger
+      add_constraint(:created_before_commit) { created_at <= commited_at }
+    end
+
+    execute """
+      CREATE OR REPLACE FUNCTION policy_versions_commit()
+        RETURNS trigger
+      LANGUAGE plpgsql AS $$
+        BEGIN
+          UPDATE policy_versions pv
+            SET commited_at = clock_timestamp()
+            WHERE commited_at IS NULL;
+          RETURN new;
+        END;
+      $$;
+
+      -- deferred constraint trigger will run on transaction commit
+      CREATE CONSTRAINT TRIGGER commit_current
+        AFTER INSERT ON policy_versions
+        INITIALLY DEFERRED
+        FOR EACH ROW
+        WHEN (NEW.commited_at IS NULL)
+        EXECUTE PROCEDURE policy_versions_commit();
+
+      -- if any version is current while creating new one, commit it
+      CREATE TRIGGER only_one_current
+        BEFORE INSERT ON policy_versions
+        FOR EACH ROW
+        EXECUTE PROCEDURE policy_versions_commit();
+
+      CREATE FUNCTION current_policy_version()
+        RETURNS SETOF policy_versions
```

Comment:
Could there really be more than one? What does it mean to return `SETOF`? When this function is called elsewhere, it seems like the caller assumes there will only be one result.

Reply:
A function returning `SETOF` returns a table of records of given type. This particular function only ever returns a table that has at most one row, but it simplifies its usage significantly that it returns a set (eg. you can use it in `FROM` clause).